### DISCUSSION TITLE
ask ppts a number between 1 and 100

### DISCUSSION THREAD

**Discussion by valentinavellani** [jspsych/jsPsych#1077] [Q&A] [answered]
Hi All,
I need to ask the participant to insert a number between 1 and 100 and to display error message if not. When the error message is displayed I need them to repeat the trial until they gave me the appropriate number. Then I want to save the chosen number only when it's between 1 and 100.

Below the trial I created to ask the ppt to insert a number:
```
  var give_number = {
            type: 'html-keyboard-response',
            stimulus: '<span style="font-size:40px;">Give me a number between 1 and 100</span>',
            choices: jsPsych.NO_KEYS,
            trial_duration: 500,
            
        };
```
Then here I want to create a function that check whether the number inserted is correct.
```
var check_number= {
            type: 'html-keyboard-response',
            stimulus: function() {
            chosen_number = jsPsych.data.getLastTrialData() // here I don't  know what to put to get the number the ppt inserted in the last triail
                if (chosen_number  != ) {   // here I need to say "if the number is not a number between 1 and 100
// then I would need to say "print a warning message" ,"ask the ppt to insert a new number", "save the new chosen number, overwriting the old wrong ones.
```
}
        };
```
Thank you in advance for your help!
Best
Valentina

**Comment by becky-gilbert** [comment, marked as answer]
I would suggest that you set this up in a slightly different way. The easiest way to loop over a set of trials for some unknown number of times is using a timeline along with a [loop function](https://www.jspsych.org/overview/timeline/#looping-timelines). 
 
And although you could use a trial with a dynamic stimulus parameter to show the "please enter a new number" message, it might be better to use a [conditional node](https://www.jspsych.org/overview/timeline/#conditional-timelines) so this message only appears if the response wasn't valid, otherwise it skips the whole trial. In the conditional function for this trial, you can (1) determine whether or not the "please enter a new number" message should appear, by checking the response from the last trial (very similar to what you've started doing already), and (2) set the value of a global variable that also tells you whether or not the loop should repeat, so that you can access the value of that global variable inside your loop function.

One issue that I noticed with your code is that you're using the html-keyboard-response plugin to get the number response, but I don't think that will work well because that plugin just captures the first (valid) key press and then ends the trial. So if someone wants to enter a two digit number, the trial will end after they enter the first digit. Instead I would suggest using the survey-text plugin. 

In order to access the response from the last trial, you had the right start with `jsPsych.data.getLastTrialData()`, you just need to get the raw data array with `.values()`, then index the first (only) data object in the array with `[0]`. Then the way that you get the response depends on how the data is saved, and thus which plugin you used for the response trial. 

If you're using the "survey-text" plugin, then the variable that holds the responses is called `responses`. The responses are stored as JSON-formatted text, so then you need to parse the JSON with `JSON.parse()`. Once you have the responses as a JavaScript object rather than JSON, you can access the question name (e.g. `Q0`) to get the response to that question. So that line would look like this:
```javascript
var last_survey_text_response = JSON.parse(jsPsych.data.getLastTrialData.values()[0].responses).Q0;
```
To check whether the response was a number between 1 and 100, you can use JavaScript's [parseInt function](https://www.w3schools.com/jsref/jsref_parseint.asp) to convert the response from a string to a number. If the participant's response can't be converted to a number then parseInt will return `NaN`, and you can check for that with JavaScript's [isNaN function](https://www.w3schools.com/jsref/jsref_isnan.asp).

Here's what it might look like when you put all of these things together:
```javascript
    var repeat_loop = false;

    var give_number = {
        type: 'survey-text',
        questions: [{prompt: '<span style="font-size:40px;">Give me a number between 1 and 100</span>', required: true, rows: 1, columns: 3}]
    };

    var repeat_message = {
        type: 'html-keyboard-response',
        stimulus: '<p style="font-size:40px;">Sorry, that was not a number between 1 and 100.</p><p style="font-size:40px;">Please press the space bar and try again.</p>',
        choices: ['space']
    };

    var repeat_message_conditional = {
        timeline: [repeat_message],
        conditional_function: function() {
            // get last response
            var response = JSON.parse(jsPsych.data.getLastTrialData().values()[0].responses).Q0;
            // convert from text to number
            var response_num = parseInt(response);
            if (isNaN(response_num) | response_num < 1 | response_num > 100) {
                // set repeat_loop to true so that the two trials in the loop timeline will repeat
                repeat_loop = true;
                // return true to show the repeat_message trial
                return true;
            } else {
                // set repeat_loop to false so that the two trials in the loop timeline will not repeat
                repeat_loop = false;
                // return false to skip the repeat_message trial
                return false;
            }
        }
    }

    var number_loop = {
        timeline: [give_number, repeat_message_conditional],
        loop_function: function() {
            // repeat this timeline if 'repeat_loop' is true
            // don't repeat this timeline if 'repeat_loop' is false
            return repeat_loop;
        }
    }

    jsPsych.init({
        timeline: [number_loop],
        on_finish: function() {
            jsPsych.data.displayData();
        }
    });
```

**Reply by valentinavellani** [reply]
Hi Becky,
thank you very much for your help!
I put the code into my Gorilla script and I added the trials in my main timeline.
What I would like to do now is to save the final first_estimate (that is when they insert a number between 1 and 100).
To do so I  inserted the line f_irst_estimate=first_estimate_firstattempt;_ in the if statements that evaluate the response given by the subject.
However this is not save in the csv file with the data.


```
var timeline = []; 
var first_estimate;


var repeat_loop = false;  // repeat look if subjs does not enter a valid number
var give_number = {
        type: 'survey-text',
        questions: [{prompt: '<span style="font-size:40px;">Give me a number between 1 and 100</span>', required: true, rows: 1, columns: 3}]
    };



  var repeat_message = {
        type: 'html-keyboard-response',
        stimulus: '<p style="font-size:40px;">Sorry, that was not a number between 1 and 100.</p><p style="font-size:40px;">Please press the space bar and try again.</p>',
        choices: ['space']
    };



var repeat_message_conditional = {
        timeline: [repeat_message],
        conditional_function: function() {
            // get last response
            response = JSON.parse(jsPsych.data.getLastTrialData().values()[0].responses).Q0;
            // convert from text to number
            first_estimate_firstattempt = parseInt(response);
            if (isNaN(first_estimate_firstattempt) ||  first_estimate_firstattempt < 1 ||  first_estimate_firstattempt > 100) {
                // set repeat_loop to true so that the two trials in the loop timeline will repeat
                repeat_loop = true;
                first_estimate=NaN
                // return true to show the repeat_message trial
                return true;
            } else {
                first_estimate=first_estimate_firstattempt;
                // set repeat_loop to false so that the two trials in the loop timeline will not repeat
                repeat_loop = false;
                // return false to skip the repeat_message trial
                return false;
            }
        }
    }



 var number_loop = {
        timeline: [give_number, repeat_message_conditional],
        loop_function: function() {
            // repeat this timeline if 'repeat_loop' is true
            // don't repeat this timeline if 'repeat_loop' is false
            return repeat_loop;
        }
    }
    
timeline.push(number_loop);
    

jsPsych.init({
		display_element: $('#gorilla')[0],
		timeline: timeline,
		on_data_update: function(data){
			gorilla.metric({
			    first_estimate:first_estimate
			
			})
		},
		on_finish: function(){
			gorilla.finish();
		}
	});
})

    

```


Thank you again!
Best,
Valentina

**Reply by becky-gilbert** [reply]
I think you have the right idea in adding the `first_estimate` data at the point where you have verified that the response is valid, which is inside your conditional function where you also end the loop and return false from the funtion. One option is to use some of [jsPsych's data functions](https://www.jspsych.org/core_library/jspsych-data/) at this point in the code to save the data. The specific functions you use will depend on how exactly you want the data to be saved. For instance, you could add the `first_estimate` data to the data for all jsPsych trials using [`jsPsych.data.addProperties`](https://www.jspsych.org/core_library/jspsych-data/#jspsychdataaddproperties). Or if you want to add the `first_estimate` data to the data from one specific trial, you could use the [filter function](https://www.jspsych.org/core_library/jspsych-data/#filter) to find the trial data that you want to add to. Then you can get the data [values](https://www.jspsych.org/core_library/jspsych-data/#values), which gives you a modifiable array of data objects, so you can add the data to a trial that way.

I don't know much about Gorilla's functions, but I see that you're using something called `gorilla.metric`, so maybe another option is to use that function at the point where you have identified the `first_estimate` value (inside your conditional function)?

**Reply by valentinavellani** [reply]
Hi Becky,
what I usually do with Gorilla, and I find it working properly so far, is to create a variable in the main script and then save it like this gorilla.metric({
			    first_estimate:first_estimate
			
			})
In Gorilla there is a separate section called "metrics" in which you manually insert the name of the variable you want to save.

However, this methods is not working here so I was thinking that maybe I placed the line in the wrong place.
As you confirmed that the place I inserted the line first_estimate=first_estimate_firstattempt; is the proper one, the issue may be related to saving data in conditional loops in Gorilla.
I will contact the Gorilla support system to see whether this is the case.


Thank you very much Becky for your kind help!
Best
Valentina

**Reply by becky-gilbert** [reply]
Ok great, thanks for the update. 

Regardless of the reason why it's not working, I don't think `on_data_update` is the right place for your `gorilla.metric` function because that function will run every time a plugin stores new data. I think you could find somewhere to put the `gorilla.metric` function so that it runs just once, if not in the conditional function then somewhere after the `number_loop`, since `first_estimate` should have a value at that point. You could try using the [call-function plugin](https://www.jspsych.org/plugins/jspsych-call-function/#jspsych-call-function) to create a 'trial' after `number_loop` that just calls the `gorilla.metric` function.

**Reply by valentinavellani** [reply]
Hi Becky,
while trying out the code I noticed that this works fine if the subject insert the following entry:
-letters and numbers (hy5u)
-numbers (65)
-letters (hyfd)
However if he insert numbers and letters (like "4hy6") the code takes it as a valid answer and does not deliver the error message
Do you have any idea about how to fix it?

Here's the code:
```
var repeat_loop = false;  // repeat look if subjs does not enter a valid number
var give_number = {
        type: 'survey-text',
        //duration: 3000,
        questions: [{prompt: '<span style="font-size:33px; position: static; top: 60px; left: 20%; text-align: center; ;">Please guess the average pollen in the garden:  </span>', required: true, rows: 1, columns: 3}],
        on_load: function() {
        
            var button = document.querySelector("#jspsych-survey-text-next");
            button.style.fontSize = "20px";
            button.style.padding = "20px";
            button.style.position = "static";
            button.style.left = "45%";
            button.style.top = "90%";
            var text_box = document.querySelector("#input-0");
            text_box.style.position = "static";
            text_box.style.left = "46%";
            text_box.style.top = "80%";
            text_box.style.fontSize = "20px";
            text_box.style.padding = "20px";
            document.querySelector("#jspsych-survey-text-form").setAttribute('autocomplete','off');
        },
        on_finish: function(data) {
                  
                  response = JSON.parse(jsPsych.data.getLastTrialData().values()[0].responses).Q0;
                  first_estimate_firstattempt= parseInt(response);
                  if (isNaN(first_estimate_firstattempt) ||  first_estimate_firstattempt < 1 ||  first_estimate_firstattempt > 100) {
                  first_estimate= NaN
                      
                  }else{
                 n_trial=n_trial+1 //INCREASE N OF TRIAL BY ONE
                  first_estimate= parseInt(response);
                  }
      
                               }

};




  var repeat_message = {
        type: 'html-keyboard-response',
        stimulus: '<p style="font-size:36px;position: static; top: 30%; left: 20%; text-align: center; ">Sorry, that was not a number between 1 and 100</p><p style="font-size:36px;position: static; top: 60%; left: 25%; text-align: center;">Please press the space bar and try again</p>',
        choices: ['space']
    };



var repeat_message_conditional = {
        timeline: [repeat_message],
        conditional_function: function() {
            // get last response
            response = JSON.parse(jsPsych.data.getLastTrialData().values()[0].responses).Q0;
            // convert from text to number
            first_estimate_firstattempt = parseInt(response);
            if (isNaN(first_estimate_firstattempt) ||  first_estimate_firstattempt < 1 ||  first_estimate_firstattempt > 100) {
                // set repeat_loop to true so that the two trials in the loop timeline will repeat
                repeat_loop = true;
                first_estimate=NaN
                // return true to show the repeat_message trial
                return true;
            } else {
                //first_estimate=parseInt(response);; THIS IS NOT SAVED PROPERLY
                // set repeat_loop to false so that the two trials in the loop timeline will not repeat
                repeat_loop = false;
                // return false to skip the repeat_message trial
                return false;
            }
        }
    }



 var number_loop = {
        timeline: [give_number, repeat_message_conditional],
        loop_function: function() {
            // repeat this timeline if 'repeat_loop' is true
            // don't repeat this timeline if 'repeat_loop' is false
            return repeat_loop;
        }
    }
    
timeline.push(number_loop);
```


Many thanks!

**Reply by becky-gilbert** [reply]
Ah, good catch! It turns out the `parseInt` function doesn't work exactly as I thought. The [parseInt docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt) say:

> If parseInt encounters a character that is not a numeral in the specified radix, it ignores it and all succeeding characters and returns the integer value parsed up to that point

So if the response _starts_ with a number, it will return the starting number(s) in the string and ignore any non-numeric characters that follow. Oops!

I googled how to determine if the _entire_ string is a number, and found [this stackoverflow answer](https://stackoverflow.com/a/175787) which explains that the `isNaN` function works on strings as well as numeric variables. So in your code, you can skip the use of `parseInt` to convert the response from string to numeric, and just use `isNaN` on the string response. 

Here's the updated code for the `give_number` trial's `on_finish` function:
```javascript
on_finish: function (data) {
  first_estimate_firstattempt = JSON.parse(jsPsych.data.getLastTrialData().values()[0].responses).Q0;
  if (isNaN(first_estimate_firstattempt) || first_estimate_firstattempt < 1 || first_estimate_firstattempt > 100) {
    first_estimate = NaN
  } else {
    n_trial = n_trial + 1 //INCREASE N OF TRIAL BY ONE
    // here we can use parseInt to convert the response from a string to number, because we've already checked that the whole string is numeric
    first_estimate = parseInt(first_estimate_firstattempt);
  }
}
```
And here's the code for the `conditional_function` in the `repeat_message_conditional`:
```javascript
conditional_function: function () {
  // get last response
  first_estimate_firstattempt = JSON.parse(jsPsych.data.getLastTrialData().values()[0].responses).Q0;
  if (isNaN(first_estimate_firstattempt) || first_estimate_firstattempt < 1 || first_estimate_firstattempt > 100) {
    // set repeat_loop to true so that the two trials in the loop timeline will repeat
    repeat_loop = true;
    first_estimate = NaN
    // return true to show the repeat_message trial
    return true;
  } else {
    //first_estimate=parseInt(response);; THIS IS NOT SAVED PROPERLY
    // set repeat_loop to false so that the two trials in the loop timeline will not repeat
    repeat_loop = false;
    // return false to skip the repeat_message trial
    return false;
  }
}
```

**Comment by valentinavellani** [comment]
Thanks Becky!
